Configuration.props

```diff
@@ -93,7 +93,7 @@
     <AndroidMxeInstallPrefix Condition=" '$(HostOS)' == 'Darwin' ">$(HostHomebrewPrefix)</AndroidMxeInstallPrefix>
     <AndroidSdkDirectory Condition=" '$(AndroidSdkDirectory)' == '' And Exists($(ANDROID_SDK_ROOT)) ">$(ANDROID_SDK_ROOT)</AndroidSdkDirectory>
     <AndroidSdkDirectory Condition=" '$(AndroidSdkDirectory)' == '' ">$(AndroidToolchainDirectory)\sdk</AndroidSdkDirectory>
-    <AndroidNdkDirectory Condition=" '$(AndroidNdkDirectory)' == '' And Exists($(ANDROID_NDK_LATEST_HOME)) ">$(ANDROID_NDK_LATEST_HOME)</AndroidNdkDirectory>
+    <AndroidNdkDirectory Condition=" '$(AndroidNdkDirectory)' == '' And Exists($(ANDROID_NDK_HOME)) ">$(ANDROID_NDK_HOME)</AndroidNdkDirectory>
```

Comment:
I'm still confused why don't we just use the `~/android-toolchain/ndk` location that we fully control? If we want to use NDK that's already installed on the machine, to avoid downloading the NDK zip, perhaps `xaprepare` should be modified to check the system location for a matching version and copy the SDK to `~/android-toolchain/ndk`, if the version matches our requirements? That way we'd remove a known unknown of the system location.

Reply:
I know that https://github.com/xamarin/xamarin-android/commit/70272dbd9600dca4b161ecd2c0251ee68d2c3481 mentions the reason to be disk space, and that `xaprepare` overwrites the system location with "our" NDK if necessary, but that appeared to have failed in the past 3 months, since we thought we were using NDK r24 (@pjcollins checked that `xaprepare` **did** overwrite the system installation with our version) but we probably didn't? Or we used a mix of whatever NDK version was installed in the system location and what `xaprepare` overlaid on it (which would explain why we found `readelf` in there).

Reply:
Unfortunately this change will not force NDK 23 usage in the tests.  This will only change where we install NDK 24, we'll now use `$(ANDROID_NDK_HOME)` rather than `$(ANDROID_NDK_LATEST_HOME)`.

We are able to write to either location declared in these variables, and we replace the contents of the path entirely with the NDK required by xaprepare assuming it is not already installed. I looked at a handful of logs and this appears to be working as expected on macOS, Windows, and Linux. I agree that we do want full control over the NDK we use for build and test regardless of system image updates. Provisioning appears to be fine in this regard, the problem comes down to how our tests resolve the NDK.

The vast majority of MSBuild tests no longer set `/p:AndroidNdkDirectory=path`, so we are left with the default NDK lookup behavior in [xamarin-android-tools](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L175). This logic was never [updated to support NDK 24](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L14), and as a result it ignores the NDK 24 that we have installed, and will use a potentially arbitrary one:

    Verbose: Skipping NDK in 'C:\Android\android-sdk\ndk\24.0.8215888': version 24.0.8215888 is out of the accepted range (major version must be between 16 and 23
    Verbose: Best NDK selected: v23.2.8568313 in C:\Android\android-sdk\ndk\23.2.8568313

Before the recent VM image update we would fall back to v22:

    Message: DEBUG  Skipping NDK in 'C:\Program Files (x86)\Android\android-sdk\ndk\23.2.8568313': version 24.0.8215888 is out of the accepted range (major version must be between 16 and 23
    Message: DEBUG  Best NDK selected: v22.1.7171670 in C:\Program Files (x86)\Android\android-sdk\ndk\22.1.7171670

On Windows, another wrench is thrown into the works. The [AotTests fixture](https://github.com/xamarin/xamarin-android/blob/745214deb03343124f5a1610434f4746ddaf669c/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AotTests.cs#L31) will resolve an NDK [with this lookup logic](https://github.com/xamarin/xamarin-android/blob/7e8cc34989bfcb5bd58cb857013ba67273c00c79/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs#L38), and then copy it to a path with special characters to test against. Unfortunately this lookup prioritizes `$(ANDROID_NDK_HOME)` over `$(ANDROID_NDK_LATEST_HOME)`, introducing a second place where the tests can use an arbitrary NDK rather than the one we install.

I think we should continue to use `$(ANDROID_NDK_LATEST_HOME)` here for performance reasons, we should not always attempt to install NDK 24 as it is now present on our test machines. Instead we can change the [NDK lookup ordering](https://github.com/xamarin/xamarin-android/blob/7e8cc34989bfcb5bd58cb857013ba67273c00c79/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs#L38) that some of the tests use to prefer `$(ANDROID_NDK_LATEST_HOME)`. We will also want to bump the [`MaximumCompatibleNDKMajorVersion`](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L14) to allow us to use the expected NDK 24 version that we provision, assuming it doesn't have other major breaking changes.

Finally, we may want to make sure to set ` /p:AndroidNdkDirectory=$(ANDROID_NDK_LATEST_HOME)` for all tests that use llvm or mkbundle if this isn't being done already.

Reply:
To summarize, I think we should return to using `$(ANDROID_NDK_LATEST_HOME)` as the default NDK path to install into or use if the version matches, and limit the options our tests have to resolve the SDK/NDK to the following:

1) Test override variables
2) Variables used by CI (`$(ANDROID_NDK_LATEST_HOME)` and `$(ANDROID_NDK_LATEST_HOME)`)
3) Local install path (`~/android-toolchain`)

```diff
diff --git a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
index 54776ab15..31c8c4f19 100644
--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
@@ -22,10 +22,6 @@ namespace Xamarin.ProjectTools
 		public static string GetAndroidSdkPath ()
 		{
 			var sdkPath = Environment.GetEnvironmentVariable ("TEST_ANDROID_SDK_PATH");
-			if (String.IsNullOrEmpty (sdkPath))
-				sdkPath = Environment.GetEnvironmentVariable ("ANDROID_SDK_PATH");
-			if (String.IsNullOrEmpty (sdkPath))
-				sdkPath = GetPathFromRegistry ("AndroidSdkDirectory");
 			if (String.IsNullOrEmpty (sdkPath))
 				sdkPath = Environment.GetEnvironmentVariable ("ANDROID_SDK_ROOT");
 			if (String.IsNullOrEmpty (sdkPath))
@@ -38,10 +34,6 @@ namespace Xamarin.ProjectTools
 		public static string GetAndroidNdkPath ()
 		{
 			var ndkPath = Environment.GetEnvironmentVariable ("TEST_ANDROID_NDK_PATH");
-			if (String.IsNullOrEmpty (ndkPath))
-				ndkPath = Environment.GetEnvironmentVariable ("ANDROID_NDK_PATH");
-			if (String.IsNullOrEmpty (ndkPath))
-				ndkPath = GetPathFromRegistry ("AndroidNdkDirectory");
 			if (String.IsNullOrEmpty (ndkPath))
 				ndkPath = Environment.GetEnvironmentVariable ("ANDROID_NDK_LATEST_HOME");
 			if (String.IsNullOrEmpty (ndkPath))
```